include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -1051,6 +1051,19 @@ struct OrtApi {
    * Prefer a value of 0 if your CPU usage is very high.
    */
   ORT_API2_STATUS(SetGlobalSpinControl, _Inout_ OrtThreadingOptions* tp_options, int allow_spinning);
+
+  /**
+   * Add a pre-allocated initializer to a session. If a model contains an initializer with a name
+   * that is same as the name passed to this API call, ORT will use this initializer instance
+   * instead of deserializing one from the model file. This is useful when you want to share
+   * the same initializer across sessions.
+   * \param name name of the initializer
+   * \param val OrtValue containing the initializer. Lifetime of 'val' and the underlying initializer buffer must be
+   * managed by the user (created using the CreateTensorWithDataAsOrtValue API) and it must outlive the session object
+   * to which it is added.
+   */
+  ORT_API2_STATUS(AddInitializer, _Inout_ OrtSessionOptions* options, _In_z_ const char* name,
+                  _In_ const OrtValue* val);
 };
```

Comment:
nit: would AddSharedInitializer be clearer?

Reply:
I went back and forth on this. Technically nothing prevents a user from using this API to provide his own initializer irrespective of whether she wants to share it with other sessions or not (although I can't think of a use case for doing this). This is why I added a comment about when calling this API would be most useful. Will leave it as-is for now.